app.go

```diff
@@ -0,0 +1,423 @@
+		return
+	}
+
+	tx, err := a.db.BeginTx(context.TODO(), nil)
+	if err != nil {
+		errored(writer, err.Error())
+		return
+	}
+	defer tx.tx.Rollback()
+
+	id, err := tx.InsertTask(rawtask)
+	if err != nil {
+		errored(writer, err.Error())
+		return
+	}
+
+	log.Info(id)
+
+	tx.tx.Commit()
+
+	url, _ := a.router.Get("getById").URL("id", id)
+	log.Info(url)
+
+	writer.Header().Set("Location", url.EscapedPath())
+	writer.WriteHeader(http.StatusCreated)
+}
+
+func (a *AsyncTasksApp) AddStatusRequest(writer http.ResponseWriter, r *http.Request) {
+	var (
+		id string
+		ok bool
+	        rawstatus AsyncTaskStatus
+		v  = mux.Vars(r)
+	)
+
+	if id, ok = v["id"]; !ok {
+		badRequest(writer, "No ID in URL")
+		return
+	}
+
+	log.Infof("Fetching async task %s", id)
+
+	tx, err := a.db.BeginTx(context.TODO(), nil)
+	if err != nil {
+		errored(writer, err.Error())
+		return
+	}
+	defer tx.tx.Rollback()
+
+	task, err := tx.GetTask(id)
+	if err != nil {
+		errored(writer, err.Error())
+		return
+	}
+
+	if task.ID == "" {
+		notFound(writer, "not found")
+		return
+	}
+
+	body, err := ioutil.ReadAll(io.LimitReader(r.Body, 104857600))
+
+	if err != nil {
+		errored(writer, err.Error())
+		return
+	}
+	if err := r.Body.Close(); err != nil {
+		errored(writer, err.Error())
+		return
+	}
+	if err := json.Unmarshal(body, &rawstatus); err != nil {
+		errored(writer, err.Error())
+		return
+	}
+
+	err = tx.InsertTaskStatus(rawstatus, id)
+	if err != nil {
+		errored(writer, err.Error())
+		return
+	}
+
+	tx.tx.Commit()
+
+	url, _ := a.router.Get("getById").URL("id", id)
+	log.Info(url)
+
+	writer.Header().Set("Location", url.EscapedPath())
+	writer.WriteHeader(http.StatusCreated)
+}
+
+func (a *AsyncTasksApp) AddBehaviorRequest(writer http.ResponseWriter, r *http.Request) {
+	var (
+		id string
+		ok bool
+	        rawbehavior AsyncTaskBehavior
+		v  = mux.Vars(r)
+	)
+
+	if id, ok = v["id"]; !ok {
+		badRequest(writer, "No ID in URL")
+		return
+	}
+
+	log.Infof("Fetching async task %s", id)
+
+	tx, err := a.db.BeginTx(context.TODO(), nil)
+	if err != nil {
+		errored(writer, err.Error())
+		return
+	}
+	defer tx.tx.Rollback()
+
+	task, err := tx.GetTask(id)
+	if err != nil {
+		errored(writer, err.Error())
+		return
+	}
+
+	if task.ID == "" {
+		notFound(writer, "not found")
+		return
+	}
+
+	body, err := ioutil.ReadAll(io.LimitReader(r.Body, 104857600))
+
+	if err != nil {
+		errored(writer, err.Error())
+		return
+	}
+	if err := r.Body.Close(); err != nil {
+		errored(writer, err.Error())
+		return
+	}
+	if err := json.Unmarshal(body, &rawbehavior); err != nil {
+		errored(writer, err.Error())
+		return
+	}
+
+	err = tx.InsertTaskBehavior(rawbehavior, id)
+	if err != nil {
+		errored(writer, err.Error())
+		return
+	}
+
+	tx.tx.Commit()
+
+	url, _ := a.router.Get("getById").URL("id", id)
+	log.Info(url)
+
+	writer.Header().Set("Location", url.EscapedPath())
+	writer.WriteHeader(http.StatusCreated)
+}
+
+type ErrorResp struct {
+	Msg string `json:"msg"`
```

Comment:
Did we want to return the standard `reason` error message (and maybe even `error_code`) our other services return? For example:
https://github.com/cyverse-de/iplant-clojure-commons/blob/d9376410d3befb246fcc4df1b6457934eff84e08/src/clojure_commons/exception.clj#L186-L187

I know this service isn't intended to be used directly by the UI, but just in case? Or even just for consistency with error logging?

Reply:
Hm, that might be better. I'm not sure what we've done in other Go services, to be honest -- I think we haven't maintained our `ERR_WHATEVER` and `reason` stuff consistently there.